base/setup/lib/utils/volutil.c

```diff
@@ -0,0 +1,207 @@
+                                 Volume->FileSystem,
+                                 sizeof(Volume->FileSystem));
+        if (!NT_SUCCESS(Status))
+            DPRINT1("InferFileSystem() failed, Status 0x%08lx\n", Status);
+    }
+    if (*Volume->FileSystem) // (!IsUnknown(Volume))
+    {
+        ASSERT(VolumeHandle);
+
+        /*
+         * Handle volume mounted with RawFS: it is
+         * either unformatted or has an unknown format.
+         */
+        if (IsUnformatted(Volume)) // (_wcsicmp(Volume->FileSystem, L"RAW") == 0)
+        {
+            /*
+             * True unformatted partitions on NT are created with their
+             * partition type set to either one of the following values,
+             * and are mounted with RawFS. This is done this way since we
+             * are assured to have FAT support, which is the only FS that
+             * uses these partition types. Therefore, having a partition
+             * mounted with RawFS and with these partition types means that
+             * the FAT FS was unable to mount it beforehand and thus the
+             * partition is unformatted.
+             * However, any partition mounted by RawFS that does NOT have
+             * any of these partition types must be considered as having
+             * an unknown format.
+             */
+            if (MbrPartitionType == PARTITION_FAT_12 ||
+                MbrPartitionType == PARTITION_FAT_16 ||
+                MbrPartitionType == PARTITION_HUGE   ||
+                MbrPartitionType == PARTITION_XINT13 ||
+                MbrPartitionType == PARTITION_FAT32  ||
+                MbrPartitionType == PARTITION_FAT32_XINT13)
+            {
+                /* The volume is unformatted */
+            }
+            else
+            {
+                /* Close the volume before dismounting */
+                NtClose(VolumeHandle);
+                VolumeHandle = NULL;
+                /*
+                 * Dismount the volume since RawFS owns it, and reset its
+                 * format (it is unknown, may or may not be actually formatted).
+                 */
+                DismountVolume(Volume);
+                *Volume->FileSystem = UNICODE_NULL;
+            }
+        }
+        /* Else, the volume is formatted */
+    }
+    /* Else, the volume has an unknown format */
+
+    /* Retrieve the volume label */
+    if (VolumeHandle)
+    {
+        IO_STATUS_BLOCK IoStatusBlock;
+        UCHAR LabelBuffer[sizeof(FILE_FS_VOLUME_INFORMATION) + 256 * sizeof(WCHAR)];
+        PFILE_FS_VOLUME_INFORMATION LabelInfo = (PFILE_FS_VOLUME_INFORMATION)LabelBuffer;
+
+        Status = NtQueryVolumeInformationFile(VolumeHandle,
+                                              &IoStatusBlock,
+                                              &LabelBuffer,
+                                              sizeof(LabelBuffer),
+                                              FileFsVolumeInformation);
+        if (NT_SUCCESS(Status))
```

Comment:
I think we can play along with the filesystem drivers (which implement the `FileFsVolumeInformation`; for [example](https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/filesystems/fastfat/volinfo.c;hb=44662eaf621860d0cf9da1b55b3a1a14e1ba0010#l349), [fastfat](https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/filesystems/fastfat/volinfo.c;hb=44662eaf621860d0cf9da1b55b3a1a14e1ba0010#l580)) and somewhat trust the VolumeLabelLength, in case they return success for the call.

Indeed, what would you check for in `VolumeLabelLength` in case of success? If it's zero, there's no label and the StringCopy operation would just NUL-terminate the output buffer (OK).

Maybe the validation check you were thinking about is checking whether that returned length is greater than the input buffer length (`LabelInfo` minus the header part)? This would mean that the driver returned success (and not `STATUS_BUFFER_OVERFLOW`) while specifying the label length had to be larger... :thinking: